### ISSUE TITLE
SGD Bug!

### ISSUE THREAD

**Issue by TangXing** [SeanNaren/CTCNetworkClient#13]
hi training will be Interrupted, if sum of gradOutput equal zore.I add gradOutput check fix this error.

```
    local function feval(x_new)
        local inputsCPU, targets = dataset:nextData()
        -- transfer over to GPU
        inputs:resize(inputsCPU:size()):copy(inputsCPU)
        gradParameters:zero()
        local predictions = self.model:forward(inputs)
        local loss = ctcCriterion:forward(predictions, targets)
        self.model:zeroGradParameters()
        local gradOutput = ctcCriterion:backward(predictions, targets)
        if  (gradOutput:sum()) ~= 0 then
            self.model:backward(inputs, gradOutput)
        end
        return loss, gradParameters
    end
```

**Comment by SeanNaren** [comment]
I think its probably a better idea as to find out why the gradOutput is 0, are you using a different dataset etc? I know when using a real system you probably don't want to add the max pooling layer (I do for the sake of reducing memory overhead). 

By adding the max pooling layer we get very close to have less time steps available than in the labels which will cause all sorts of issues!

**Comment by TangXing** [comment]
Yes， i using a different dataset，This code speed up 10 times than i used before.I will do more test on my dataset.

**Comment by SeanNaren** [comment]
The code speeds up using a different dataset? May I ask which dataset you are using?

**Comment by TangXing** [comment]
Oh  i means ,the code speeds up than eesen which i used before.The data set i use is  a personal data set for Mandarin speech recognize.

**Comment by SeanNaren** [comment]
Ahh I see. Well removing the max pooling layer should hopefully prevent the gradOutputs from going to 0, in the DeepSpeech network (as I've read in the paper) it doesn't have max poolings. Should hopefully fix the grad outputs from being 0s!

**Comment by minushuang** [comment]
Hi @TangXing , some questions when I am using this code to train a  Mandarin model, can i have your wechat and ask for some advice? thanks!
